packages/yew/src/functional/hooks/use_effect.rs

```diff
@@ -1,9 +1,59 @@
-use crate::functional::use_hook;
+use crate::functional::{hook, use_hook};
 use std::rc::Rc;
 
-struct UseEffect<Destructor> {
-    runner: Option<Box<dyn FnOnce() -> Destructor>>,
-    destructor: Option<Box<Destructor>>,
+type UseEffectRunnerFn<T, D> = Box<dyn FnOnce(&T) -> D>;
+
+struct UseEffectBase<T, D> {
+    runner_with_deps: Option<(Rc<T>, UseEffectRunnerFn<T, D>)>,
+    destructor: Option<Box<D>>,
+    deps: Option<Rc<T>>,
+}
+
+#[hook]
+fn use_effect_base<T, D, R>(callback: impl FnOnce(&T) -> D + 'static, deps: T, effect_changed_fn: R)
+where
+    T: 'static,
+    D: FnOnce() + 'static,
+    R: FnOnce(Option<&T>, Option<&T>) -> bool + 'static,
```

Comment:
I would prefer staying clear on what the generics for any onlooker that might inspect the inner works of yew.
```suggestion
    DESTRUCTOR: FnOnce() + 'static,
    RUNNER: FnOnce(Option<&T>, Option<&T>) -> bool + 'static,
```
or
```suggestion
    Destructor: FnOnce() + 'static,
    Runner: FnOnce(Option<&T>, Option<&T>) -> bool + 'static,
```

Reply:
This is actually a convention throughout Rust's standard library.

https://doc.rust-lang.org/std/iter/trait.Iterator.html

Reply:
@futursolo In our case the generics a a bit more complicated. std just usually consists of only like 2 basic generic.
I would like to avoid problems like actix has where you get 6 different generics and have no clue what is what.

I really think more clarity wont hurt in our case.

Its fine if you don't change this, unless other maintainers agree with me.

Reply:
I have reduced the number of generics to the same level as the standard library.